infrastructure/src/main/java/org/corfudb/infrastructure/CorfuMsgHandler.java

```diff
@@ -75,95 +98,151 @@ public CorfuMsgHandler() {
+            throw new UnrecoverableCorfuError("Handler for " + annotation.type()
+                + " already registered!");
+        }
+        // convert the method into a Java8 Lambda for maximum execution speed...
+        try {
+            Handler<CorfuMsg> h;
+            if (Modifier.isStatic(method.getModifiers())) {
+                MethodHandle mh = caller.unreflect(method);
+                MethodType mt = mh.type().changeParameterType(0, CorfuMsg.class);
+                h = (Handler<CorfuMsg>) LambdaMetafactory.metafactory(caller,
+                    "handle", MethodType.methodType(Handler.class),
+                    mt, mh, mh.type())
+                    .getTarget().invokeExact();
+
+            } else {
+                // instance method, so we need to capture the type.
+                MethodType mt = MethodType.methodType(method.getReturnType(),
+                        method.getParameterTypes());
+                MethodHandle mh = caller.findVirtual(server.getClass(), method.getName(), mt);
+                MethodType mtGeneric = mh.type().changeParameterType(1, CorfuMsg.class);
+                h = (Handler<CorfuMsg>) LambdaMetafactory.metafactory(caller,
+                    "handle",
+                    MethodType.methodType(Handler.class, server.getClass()),
+                    mtGeneric.dropParameterTypes(0, 1), mh,
+                    mh.type().dropParameterTypes(0, 1)).getTarget()
+                    .bindTo(server).invoke();
+            }
+            // Install pre-conditions on handler
+            final Handler<CorfuMsg> handler =
+                    generateConditionalHandler(server, annotation.type(), h);
+            // Install the handler in the map
+            handlerMap.put(annotation.type(), handler);
+        } catch (Throwable e) {
+            log.error("Exception during message handler registration", e);
+            throw new UnrecoverableCorfuError(e);
+
+        }
     }
 
+    /** Generate a conditional handler, which instruments the handler with metrics if enabled
+     *  and checks whether the server is shutdown and ready.
+     * @param server        The {@link AbstractServer} the message is being handled for.
+     * @param type          The {@link CorfuMsgType} the message is being handled for.
+     * @param handler       The {@link Handler} which handles the message.
+     * @return              A new {@link Handler} which conditionally executes the handler
+     *                      based on preconditions (whether the server is shutdown/ready).
+     */
+    private Handler<CorfuMsg> generateConditionalHandler(@Nonnull final AbstractServer server,
+            @Nonnull final CorfuMsgType type,
+            @Nonnull final Handler<CorfuMsg> handler) {
+        if (!MetricsUtils.isMetricsCollectionEnabled()) {
+            // If metrics are disabled, register the handler without instrumentation.
+            return (msg, ctx, r) -> {
+                if (server.isShutdown()) {
```

Comment:
I think this logic should go somewhere more common and visible. It should be used in generation but if possible can we keep it in a separate method of it's own.